File: codeforce/exam/l.py
maxFlow = 0
startSum = 0
targetSum = 0

class Edge:
    def __init__(self, from_vertex=-1, to_vertex=-1, flow=0):
        self.from_vertex = from_vertex
        self.to_vertex = to_vertex
        self.flow = flow

class Graph:
    def __init__(self, height, width):
        self.height = height
        self.width = width
        self.v_num = height * width + 2
        self.checked = [False] * self.v_num
        self.edge = [[] for _ in range(self.v_num)]

    def add_edge(self, from_vertex, to_vertex, flow=1):
        self.edge[from_vertex].append(Edge(from_vertex, to_vertex, flow))
        self.edge[to_vertex].append(Edge(to_vertex, from_vertex, 0))  # Добавляем обратное ребро

    def uncheck_all(self):
        self.checked = [False] * self.v_num

    def to_vert_num(self, i, j):
        return i * self.width + j + 1

    def init(self):
        global startSum, targetSum

        for i in range(self.height):
            row = input().strip()
            for j in range(self.width):
                atom = row[j]
                valence_value = valence(atom)
                if (i % 2 + j % 2) % 2 == 0:
                    self.add_edge(0, self.to_vert_num(i, j), valence_value)
                    startSum += valence_value
                    if i + 1 < self.height:
                        self.add_edge(self.to_vert_num(i, j), self.to_vert_num(i + 1, j))
                    if j + 1 < self.width:
                        self.add_edge(self.to_vert_num(i, j), self.to_vert_num(i, j + 1))
                    if i - 1 >= 0:
                        self.add_edge(self.to_vert_num(i, j), self.to_vert_num(i - 1, j))
                    if j - 1 >= 0:
                        self.add_edge(self.to_vert_num(i, j), self.to_vert_num(i, j - 1))
                else:
                    self.add_edge(self.to_vert_num(i, j), self.v_num - 1, valence_value)
                    targetSum += valence_value

def find_back_edge(G, edge):
    for child in range(len(G.edge[edge.to_vertex])):
        if G.edge[edge.to_vertex][child].to_vertex == edge.from_vertex:
            return child
    return -1

def valence(atom):
    return {'H': 1, 'O': 2, 'N': 3, 'C': 4}.get(atom, 0)

def try_push_flow(G, current=0, capacity=float('inf')):
    global maxFlow

    G.checked[current] = True
    if current == G.v_num - 1:
        maxFlow += capacity
        return capacity

    for e in range(len(G.edge[current])):
        edge = G.edge[current][e]

        if G.checked[edge.to_vertex] or edge.flow == 0:
            continue

        min_capacity = try_push_flow(G, edge.to_vertex, min(edge.flow, capacity))
        if min_capacity > 0:
            G.edge[current][e].flow -= min_capacity
            back_edge_num = find_back_edge(G, edge)
            if back_edge_num != -1:
                G.edge[edge.to_vertex][back_edge_num].flow += min_capacity
            else:
                G.edge[edge.to_vertex].append(Edge(edge.to_vertex, current, min_capacity))
            return min_capacity

    return 0

if __name__ == "__main__":
    height, width = map(int, input().strip().split())
    maxFlow = 0
    startSum = 0
    targetSum = 0

    G = Graph(height, width)
    G.init()

    while try_push_flow(G):
        G.uncheck_all()

    if maxFlow == startSum == targetSum != 0:
        print("Valid")
    else:
        print("Invalid")

File: codeforce/exam/l.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Edge struct {
	from int
	to   int
	flow int
}

type Graph struct {
	vNum    int
	height  int
	width   int
	checked []bool
	edge    [][]Edge
}

var maxFlow = 0
var startSum = 0
var targetSum = 0

func main() {
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	var height, width int
	fmt.Sscanf(input, "%d %d", &height, &width)

	G := NewGraph(height, width)
	G.Init(height, width)

	for tryPushFlow(G, 0, 5) > 0 {
		G.UncheckAll()
	}

	if maxFlow == startSum && maxFlow == targetSum && maxFlow != 0 {
		fmt.Println("Valid")
	} else {
		fmt.Println("Invalid")
	}
}

func NewGraph(height, width int) *Graph {
	G := &Graph{
		height:  height,
		width:   width,
		vNum:    height*width + 2,
		checked: make([]bool, height*width+2),
		edge:    make([][]Edge, height*width+2),
	}
	return G
}

func (G *Graph) AddEdge(from, to, flow int) {
	G.edge[from] = append(G.edge[from], Edge{from, to, flow})
}

func (G *Graph) UncheckAll() {
	for i := range G.checked {
		G.checked[i] = false
	}
}

func (G *Graph) ToVertNum(i, j int) int {
	return i*G.width + j + 1
}

func (G *Graph) Init(height, width int) {
	reader := bufio.NewReader(os.Stdin)
	for i := 0; i < height; i++ {
		for j := 0; j < width; j++ {
			atom, _ := reader.ReadString(' ')
			atom = strings.TrimSpace(atom)
			val := valence(atom[0])
			if (i%2+j%2)%2 == 0 {
				G.AddEdge(0, G.ToVertNum(i, j), val)
				startSum += val
				if i+1 < height {
					G.AddEdge(G.ToVertNum(i, j), G.ToVertNum(i+1, j), 1)
				}
				if j+1 < width {
					G.AddEdge(G.ToVertNum(i, j), G.ToVertNum(i, j+1), 1)
				}
				if i-1 >= 0 {
					G.AddEdge(G.ToVertNum(i, j), G.ToVertNum(i-1, j), 1)
				}
				if j-1 >= 0 {
					G.AddEdge(G.ToVertNum(i, j), G.ToVertNum(i, j-1), 1)
				}
			} else {
				G.AddEdge(G.ToVertNum(i, j), G.vNum-1, val)
				targetSum += val
			}
		}
	}
}

func valence(atom byte) int {
	switch atom {
	case 'H':
		return 1
	case 'O':
		return 2
	case 'N':
		return 3
	case 'C':
		return 4
	}
	return 0
}

func findBackEdge(G *Graph, edge Edge) int {
	for idx, e := range G.edge[edge.to] {
		if e.to == edge.from {
			return idx
		}
	}
	return -1
}

func tryPushFlow(G *Graph, current, capacity int) int {
	G.checked[current] = true
	if current == G.vNum-1 {
		maxFlow += capacity
		return capacity
	}
	for i := range G.edge[current] {
		edge := G.edge[current][i]
		if G.checked[edge.to] || edge.flow == 0 {
			continue
		}
		minCapacity := tryPushFlow(G, edge.to, min(edge.flow, capacity))
		if minCapacity > 0 {
			G.edge[current][i].flow -= minCapacity
			backEdgeNum := findBackEdge(G, edge)
			if backEdgeNum != -1 {
				G.edge[edge.to][backEdgeNum].flow += minCapacity
			} else {
				G.edge[edge.to] = append(G.edge[edge.to], Edge{edge.to, edge.from, minCapacity})
			}
			return minCapacity
		}
	}
	return 0
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
